xla/service/gpu/buffer_allocations.cc

```diff
@@ -19,27 +19,31 @@ limitations under the License.
 #include <set>
 
 #include "xla/service/buffer_assignment.h"
+#include "xla/service/gpu/gpu_constants.h"
 #include "xla/status.h"
 #include "xla/stream_executor/device_memory.h"
 
 namespace xla {
 namespace gpu {
 
 Status BufferAllocations::TearDown(
-    const std::set<se::DeviceMemoryBase>& live_addresses,
+    const absl::flat_hash_set<BufferAllocation::Index>& live_allocations,
     absl::Span<const BufferAllocation> allocations) {
   // Deallocate temporary buffers, taking care to try to deallocate all of them
   // even if one of the deallocations fails.
   Status status;
   const int64_t num_buffers = allocations.size();
   for (BufferAllocation::Index i = 0; i < num_buffers; ++i) {
     const BufferAllocation& allocation = allocations[i];
-    se::DeviceMemoryBase buffer_address = GetDeviceAddress(allocation.index());
     // Deallocate buffers marked "maybe_live_out" but aren't actually live out,
-    // and temp buffers.
-    if ((allocation.maybe_live_out() &&
-         !live_addresses.count(buffer_address)) ||
-        allocation.IsPreallocatedTempBuffer()) {
+    // and temp buffers. For external allocations, the free should performed by
+    // external allocator, so skip external allocations free here.
+    if (((allocation.maybe_live_out() &&
```

Comment:
With memory spaces and per-memory-space-allocators this should me much simpler. The tricky part is that VA allocated buffers can't be live-out, but this should be controlled by a memory assingment pass